llvm/lib/MC/WasmObjectWriter.cpp

```diff
@@ -1769,6 +1769,11 @@ uint64_t WasmObjectWriter::writeOneObject(MCAssembler &Asm,
       WS.setIndex(InvalidIndex);
       continue;
     }
+    // Contents of .init_array sections are handled elsewhere.
+    if (WS.isDefined() &&
+        WS.getSection().getName().starts_with(".init_array")) {
```

Comment:
How is it possible to define a symbol within this section?  Can you give and example?  It seems like it should be just a sequence of pointers, but I guess somehow you can have symbol that points into that sequence?

Reply:
Yes, the following C code does so:

```
#import <stdio.h>

void init1() { puts("init1"); }
__attribute__((section(".init_array"))) void (*p_init[3])(void) = { &init1, &init1, &init1 };

int main() { return 0; }
```

Under (e.g.) Linux, a list of pointers is written to the `.init_array` section and a symbol `p_init` points to the sequence.

```
$ clang -c symbol.c -o symbol.o
$ obj2yaml symbol.o
[...]
Symbols:
[...]
  - Name:            p_init
    Type:            STT_OBJECT
    Section:         .init_array
    Binding:         STB_GLOBAL
    Size:            0x18
```

However, with a WebAssembly object this does not make as much sense. `WasmObjectWriter.cpp` converts the sequence of pointers in `.init_array` section into entries in the `WASM_INIT_FUNC` subsection of the custom `linking` section instead. There they are no longer written as a list of pointers, but indexes into a table.

So, this particular change makes it so that the `p_init` symbol is not written into the WebAssembly object file. In the current version a symbol is actually written, but it is broken in any case since it points to the incorrect section (leading to the `invalid data segment index: 0` bug shown in the OP).

Reply:
Should we error out int this maybe?   It seems like if somebody had a usage of `p_init` somewhere it would simply not work.  How about something like  "symbols within `.init_array` sections are not supported"?

Reply:
> Should we error out int this maybe? It seems like if somebody had a usage of `p_init` somewhere it would simply not work. How about something like "symbols within .init_array sections are not supported"?

A reasonable argument, but I think it would get in the way.

It seems to me the `p_init` symbol is just a side-effect of this pattern in C, where the real intention is just to add the list of function pointers to the `.init_array` section. We don't actually need to reference `p_init` anywhere, but if we throw an error here we unfortunately break the pattern.

Ultimately what I am interested in is some Rust code that uses a [similar pattern](https://github.com/dtolnay/inventory/blob/bb269d9bf2d464a1c6862aba6222f82c325e7c4b/src/lib.rs#L432-L462). Essentially:

```rust
unsafe extern "C" fn __ctor() {
    [...]
}
#[cfg_attr(link_section = ".init_array")]
static __CTOR: unsafe extern "C" fn() = __ctor;
```

So I would be grateful if we could continue to allow for this use case and not error out, even if the side-effect symbol cannot be referenced.

Another option is to keep the current situation, but it means somehow otherwise handling the `invalid data segment index` error when using `obj2yaml` in the unit test for this PR. I think by avoiding emitting the symbols we are doing no worse than before, and perhaps a little better in that regard.

Reply:
What will happen though if there is a reference to the `p_init` symbol though, either within the object itself, or in some other object.   Would this result in a relocation with against a non-existent symbol?

Reply:
What do you think about landing this PR without this part of the patch and then following up with another PR for e.g. "Handle symbols in init_array sections".

IIUC the test in this PR doesn't contain such symbols anyway so we would likely want a separate test anyway?

Reply:
> ```rust
> unsafe extern "C" fn __ctor() {
>     [...]
> }
> #[cfg_attr(link_section = ".init_array")]
> static __CTOR: unsafe extern "C" fn() = __ctor;
> ```

In this example which is the symbol that lives in the `.init_array` section?  I guess it would be `__CTOR`?

Reply:
> What will happen though if there is a reference to the p_init symbol though, either within the object itself, or in some other object. Would this result in a relocation with against a non-existent symbol?

Looks like in both cases you get an error at link time:
```
wasm-ld: error: symbol.o: invalid data relocation: init1
```

I see, that's probably not ideal.

> In this example which is the symbol that lives in the `.init_array` section? I guess it would be __CTOR?

Yes. Though Rust mangles the name.

> What do you think about landing this PR without this part of the patch and then following up with another PR for e.g. "Handle symbols in init_array sections".

I think I'm OK with that, a fresh set of eyes on it in a new PR might be good in any case. I'll push a commit to remove this part of the PR.